osquery/events/pathset.h

```diff
@@ -11,149 +11,142 @@
 #pragma once
 
 #include <mutex>
-#include <set>
 #include <string>
 #include <vector>
 
 #include <boost/noncopyable.hpp>
 #include <boost/tokenizer.hpp>
+#include <boost/utility/string_view.hpp>
 
 #include <osquery/core.h>
 #include <osquery/filesystem.h>
 
 namespace osquery {
 
 /**
- * @brief multiset based implementation for path search.
+ * @brief Vector based implemention for path search.
  *
- * 'multiset' is used because with patterns we can serach for equivalent keys.
- * Since  '/This/Path/is' ~= '/This/Path/%' ~= '/This/Path/%%' (equivalent).
+ * PathSet applies the following policy -
+ * patternedPath -    Path can contain pattern '%' and '%%'.
+ *                    Path components containing partial patterns are also
+ *                    supported e.g. '/This/Path/xyz%' or '/This/Path/%xyz'
+ *                    or '/This/Path/%xyz%' but '/This/Path/xy%z'
+ *                    is not supported i.e. 'xy%z' is considered as normal
+ *                    string.
  *
- * multiset is protected by lock. It is threadsafe.
- *
- * PathSet can take any of the two policies -
- * 1. patternedPath - Path can contain pattern '%' and '%%'.
- *                    Path components containing only '%' and '%%' are supported
- *                    e.g. '/This/Path/%'.
- *                    Path components containing partial patterns are not
- *                    supported e.g. '/This/Path/xyz%' ('xyz%' will not be
- *                    treated as pattern).
  */
 template <typename PathType>
 class PathSet : private boost::noncopyable {
  public:
   void insert(const std::string& str) {
     auto pattern = str;
     replaceGlobWildcards(pattern);
```

Comment:
In case of the 'file_paths_query' and 'file_paths', file_paths parser does this while parsing. Is there any reason for having exception here?
https://github.com/facebook/osquery/blob/master/osquery/config/parsers/file_paths.cpp#L99
https://github.com/facebook/osquery/blob/master/osquery/config/parsers/file_paths.cpp#L129

Reply:
This is a utility class. It should take care of this.

Reply:
I think, point is that we should keep this classes 'pure' in terms of not to be dependent on the filesystem(replaceGlobWildcards is depended). It makes class easier to test, and also makes more sense in the context I mentioned earlier.

Reply:
I understand your point. I think we could benefit more by making this class more testable and assimilating it better into our config parsers.